modin/pandas/test/test_groupby.py

```diff
@@ -1714,3 +1766,110 @@ def test_not_str_by(by, as_index):
+@pytest.mark.parametrize("external_by_length", [0, 1, 2])
+@pytest.mark.parametrize("has_categorical_by", [True, False])
+@pytest.mark.parametrize(
+    "agg_func",
+    [
+        pytest.param(
+            lambda grp: grp.apply(lambda df: df.dtypes), id="modin_dtype_impl"
+        ),
+        pytest.param(lambda grp: grp.apply(lambda df: df.sum()), id="apply(sum)"),
+        pytest.param(lambda grp: grp.count(), id="count"),
+        pytest.param(lambda grp: grp.nunique(), id="nunique"),
+        pytest.param(
+            {"c": "sum", "d": "nunique"}, id="dict_agg_no_intersection_with_by"
+        ),
+        pytest.param(
+            {"b": "mean", "c": "sum", "d": "nunique"},
+            id="dict_agg_has_intersection_with_by",
+        ),
+        pytest.param(
+            {"a": "nunique", "c": "sum", "d": "nunique"},
+            id="dict_agg_has_intersection_with_categorical_by",
+        ),
+    ],
+)
+def test_handle_as_index(
+    internal_by_length, external_by_length, has_categorical_by, agg_func, request
+):
+    """
+    Test ``GroupBy.handle_as_index``.
+
+    The role of the ``handle_as_index`` method is to build a groupby result considering
+    ``as_index=False`` from the result that was computed with ``as_index=True``.
+
+    So the testing flow is the following:
+        1. Compute GroupBy result with the ``as_index=True`` parameter via Modin.
+        2. Build ``as_index=False`` result from the ``as_index=True`` using ``handle_as_index`` method.
+        3. Compute GroupBy result with the ``as_index=False`` parameter via Pandas (is the reference result).
+        4. Compare the result from the second step with the reference.
+    """
+    by_length = sum((internal_by_length, external_by_length))
+    if by_length == 0:
+        return
+
+    if (
+        has_categorical_by
+        and by_length > 1
+        and (
+            isinstance(agg_func, dict)
+            or ("nunique" in request.node.callspec.id.split("-"))
+        )
+    ):
+        pytest.skip(
+            "The linked bug makes pandas raise an exception when 'by' is categorical: "
+            "https://github.com/pandas-dev/pandas/issues/36698"
+        )
+
+    data = {
+        "a": [1, 1, 2, 2],
+        "b": [2, 2, 3, 3],
+        "c": [1, 2, 3, 4],
+        "d": [5, 6, 7, 8],
+    }
+
+    if isinstance(agg_func, dict):
+        selection = list(agg_func.keys())
+        agg_dict = agg_func
+        agg_func = lambda grp: grp.agg(agg_dict)  # noqa: E731 (lambda assignment)
+    else:
+        selection = None
+
+    df = pandas.DataFrame(data)
+    external_by_cols = GroupBy.validate_by(df.add_prefix("external_"))
+
+    if has_categorical_by:
+        df = df.astype({"a": "category"})
+
+    internal_by = df.columns[:internal_by_length].tolist()
+    external_by = external_by_cols[:external_by_length]
+
+    by = internal_by + external_by
+
+    grp_res = pd.DataFrame(df).groupby(by, as_index=True)
+    grp_ref = df.groupby(by, as_index=False)
```

Comment:
please don't save a few typed letters and spell the words in the full - I suggest replacing `res` with `result` and `ref` with `reference` or smth similar (here and in other places of this function)